aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -0,0 +1,474 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::task::Transaction;
+use anyhow::bail;
+use aptos_mvhashmap::{
+    types::{MVDataError, MVDataOutput, TxnIndex, Version},
+    versioned_data::VersionedData,
+    versioned_group_data::VersionedGroupData,
+};
+use aptos_types::{state_store::state_value::StateValueMetadataKind, write_set::TransactionWrite};
+use derivative::Derivative;
+use std::{
+    collections::{
+        hash_map::{
+            Entry,
+            Entry::{Occupied, Vacant},
+        },
+        HashMap,
+    },
+    sync::Arc,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) enum ReadKind {
+    Exists,
+    Metadata,
+    Value,
+}
+
+/// The enum captures the state that the transaction execution extracted from
+/// a read callback to block executor, in order to be validated by Block-STM.
+/// The captured state is fine-grained, e.g. it distinguishes between reading
+/// a full value, which is compared conservatively based on the version (i.e.
```

Comment:
Oh, very good point, we could literally do it right here. I think we wouldn't even need a different type, because this type has Arc already, so we decide how we derive the comparison. Currently we compare versions, but we can see if the length inside the Arc is small and compare them instead. Before it was larger change, now it would be probably 4 lines. First, this makes me happy because usually a sign of a correct design direction. Second, should I do this already? The gains, irresistible..
Or a separate PR/commit?

Reply:
Separate :)

Reply:
added a TODO.